src/api_manager/context/request_context_test.cc

```diff
@@ -0,0 +1,387 @@
+};
+
+class MockRequest : public Request {
+ public:
+  MockRequest(const std::string& client_ip,
+              const std::map<std::string, std::string>& header)
+      : client_ip_(client_ip),
+        header_(header) {
+  }
+
+  virtual ~MockRequest() {
+  }
+
+  bool FindHeader(const std::string &name, std::string *header) override {
+    auto it = header_.find(name);
+    if (it != header_.end()) {
+      header->assign(it->second);
+      return true;
+    }
+
+    return false;
+  }
+
+  std::string GetClientIP() {
+    return client_ip_;
+  }
+
+  std::string GetRequestHTTPMethod() override {
+    return "GET";
+  }
+
+  std::string GetQueryParameters() override {
+    return "";
+  }
+
+  std::string GetUnparsedRequestPath() override {
+    return "/echo";
+  }
+
+  bool FindQuery(const std::string &name, std::string *query) override {
+    return false;
+  }
+
+  MOCK_METHOD2(AddHeaderToBackend,
+      utils::Status(const std::string &, const std::string &));
+  MOCK_METHOD1(SetAuthToken, void(const std::string &));
+  MOCK_METHOD0(GetFrontendProtocol,
+      ::google::api_manager::protocol::Protocol());
+  MOCK_METHOD0(GetBackendProtocol, ::google::api_manager::protocol::Protocol());
+  MOCK_METHOD0(GetRequestPath, std::string());
+  MOCK_METHOD0(GetInsecureCallerID, std::string());
+  MOCK_METHOD0(GetRequestHeaders, std::multimap<std::string, std::string> *());
+  MOCK_METHOD0(GetGrpcRequestBytes, int64_t());
+  MOCK_METHOD0(GetGrpcResponseBytes, int64_t());
+  MOCK_METHOD0(GetGrpcRequestMessageCounts, int64_t());
+  MOCK_METHOD0(GetGrpcResponseMessageCounts, int64_t());
+private:
+  const std::string client_ip_;
+  const std::map<std::string, std::string> header_;
+};
+
+}
+
+class RequestContextTest : public ::testing::Test {
+ protected:
+RequestContextTest()
+    : callback_run_count_(0) {
+}
+
+void SetUp() {
+  callback_run_count_ = 0;
+  call_history_.clear();
+}
+
+ protected:
+std::vector<std::string> call_history_;
+int callback_run_count_;
+};
+
+TEST_F(RequestContextTest, ClientIPAddressNoOverrideTest) {
+std::unique_ptr<ApiManagerEnvInterface> env(new MockApiManagerEnvironment());
+
+std::shared_ptr<context::GlobalContext> global_context(
+    new context::GlobalContext(
+        std::move(env), std::string(kServerConfigWithoutClientIPExperiment)));
+
+std::unique_ptr<Config> config = Config::Create(global_context->env(),
+                                                std::string(kServiceConfig1));
+
+std::shared_ptr<context::ServiceContext> service_context(
+    new context::ServiceContext(global_context, std::move(config)));
+
+std::unique_ptr<MockRequest> request(
+    new MockRequest("4.4.4.4", {
+                        {"X-Forwarded-For", "1.1.1.1, 2.2.2.2, 3.3.3.3"}, {
+                            "apiKey", "test-api-key"}}));
+
+RequestContext context(service_context, std::move(request));
+
+service_control::CheckRequestInfo info;
+
+context.FillCheckRequestInfo(&info);
+
+EXPECT_EQ("4.4.4.4", info.client_ip);
```

Comment:
could you format this file?  it is very hard to read

Reply:
Fixed formatting